File: poolvr/physics/poly_solvers.c
#include <math.h>
#include <complex.h>


#define PIx2 (M_PI * 2.0)
#define ZERO_TOLERANCE 1e-9
#define IMAG_TOLERANCE 1e-8
#define IMAG_TOLERANCE_SQRD (IMAG_TOLERANCE * IMAG_TOLERANCE)


double complex CUBE_ROOTS_OF_1[] = {1.0, cexp(I*PIx2/3), cexp(I*2*PIx2/3)};


void quartic_solve(double P[5], double complex out[4]) {
  double e = P[0] / P[4];
  double d = P[1] / P[4];
  double c = P[2] / P[4];
  double b = P[3] / P[4];
  double bb = b*b;
  double p = 0.125 * (8*c - 3*bb);
  double q = 0.125 * (bb*b - 4*b*c + 8*d);
  double r = (-3*bb*bb + 256*e - 64*b*d + 16*bb*c) / 256;
  c = p;
  d = q;
  e = r;
  double cc = c*c;
  double dd = d*d;
  double ee = e*e;
  double ccc = cc*c;
  double Delta = -27*dd*dd - 128*cc*ee + 256*ee*e + 16*cc*cc*e + 144*c*dd*e - 4*ccc*dd;
  double D = 64*e - 16*cc;
  double Delta_0 = cc + 12*e;
  double Delta_1 = 2*ccc + 27*dd - 72*c*e;
  double complex S;
  double complex SSx4;
  if (Delta_1 > 0 && p < 0 && D < 0) {
    double complex phi = cacos(Delta_1 / (2*csqrt(Delta_0*Delta_0*Delta_0)));
    S = 0.5 * csqrt((-2*p + 2*csqrt(Delta_0)*ccos(phi/3))/3);
    SSx4 = 4*S*S;
  } else {
    double complex Q = cpow(0.5*(Delta_1 + csqrt(-27*Delta)), 1.0/3);
    if (Delta != 0) {
      double abs_SSx4_max = 0;
      double abs_SSx4;
      double complex SSx4_max;
      for (int ir = 0; ir < 3; ir++) {
	SSx4 = (-2.0*p + (Q*CUBE_ROOTS_OF_1[ir] + Delta_0/(Q*CUBE_ROOTS_OF_1[ir]))) / 3;
	abs_SSx4 = cabs(SSx4);
	if (abs_SSx4 > abs_SSx4_max) {
	  abs_SSx4_max = abs_SSx4;
	  SSx4_max = SSx4;
	}
      }
      SSx4 = SSx4_max;
    } else {
      SSx4 = (-2.0*p + (Q + Delta_0/Q)) / 3;
    }
    S = 0.5*csqrt(SSx4);
  }
  double complex sqrtp = csqrt(-SSx4 - 2*p + q/S);
  double complex sqrtm = csqrt(-SSx4 - 2*p - q/S);
  out[0] = -0.25*b - S + 0.5*sqrtp;
  out[1] = -0.25*b - S - 0.5*sqrtp;
  out[2] = -0.25*b + S + 0.5*sqrtm;
  out[3] = -0.25*b + S - 0.5*sqrtm;
}


int sort_complex_conjugate_pairs(double complex roots[4]) {
  int npairs = 0;
  double complex r;
  double complex r_conj;
  for (int i = 0; i < 3; i++) {
    r = roots[i];
    if (fabs(cimag(r)) > IMAG_TOLERANCE) {
      for (int j = 1; j + i < 4; j++) {
	r_conj = roots[i+j];
	if (fabs(creal(r) - creal(r_conj)) < ZERO_TOLERANCE
	    && fabs(cimag(r) + cimag(r_conj)) < ZERO_TOLERANCE) {
	  roots[i] = roots[2*npairs];
	  roots[i+j] = roots[2*npairs+1];
	  roots[2*npairs] = r;
	  roots[2*npairs+1] = r_conj;
	  npairs++;
	  i++;
	  break;
	}
      }
    }
  }
  return npairs;
}


double find_min_quartic_root_in_real_interval(double P[5], double t0, double t1) {
  double complex roots[4];
  quartic_solve(P, roots);
  int npairs = sort_complex_conjugate_pairs(roots);
  double min_root = NAN;
  for (int i = 2*npairs; i < 4; i++) {
    double complex r = roots[i];
    if (t0 < creal(r) && creal(r) < t1 &&
	cimag(r)*cimag(r) / (creal(r)*creal(r) + cimag(r)*cimag(r))
	< IMAG_TOLERANCE_SQRD) {
      min_root = t1 = creal(r);
    }
  }
  return min_root;
}


double find_collision_time(double a_i[3][3], double a_j[3][3],
			   double R, double t0, double t1) {
  double a_ji[3][3] = {{a_i[0][0] - a_j[0][0],
			a_i[0][1] - a_j[0][1],
			a_i[0][2] - a_j[0][2]},
		       {a_i[1][0] - a_j[1][0],
			a_i[1][1] - a_j[1][1],
			a_i[1][2] - a_j[1][2]},
		       {a_i[2][0] - a_j[2][0],
			a_i[2][1] - a_j[2][1],
			a_i[2][2] - a_j[2][2]}};
  double a_x = a_ji[2][0];
  double a_y = a_ji[2][2];
  double b_x = a_ji[1][0];
  double b_y = a_ji[1][2];
  double c_x = a_ji[0][0];
  double c_y = a_ji[0][2];
  double p[5];
  p[4] = a_x*a_x + a_y*a_y;
  p[3] = 2 * (a_x*b_x + a_y*b_y);
  p[2] = b_x*b_x + b_y*b_y + 2 * (a_x*c_x + a_y*c_y);
  p[1] = 2 * (b_x*c_x + b_y*c_y);
  p[0] = c_x*c_x + c_y*c_y - 4 * R*R;
  return find_min_quartic_root_in_real_interval(p, t0, t1);
}

File: poolvr/physics/poly_solvers.f90
MODULE poly_solvers
  USE iso_c_binding, only: c_double, c_double_complex, c_int
  IMPLICIT NONE

  real(c_double), parameter :: ZERO_TOLERANCE = 1.d-14
  real(c_double), parameter :: IMAG_TOLERANCE = 5.d-8
  real(c_double), parameter :: IMAG_TOLERANCE_SQRD = IMAG_TOLERANCE**2
  real(c_double), parameter :: PI = acos(-1.d0)
  real(c_double), parameter :: PIx2 = 2*PI
  complex(c_double_complex), dimension(3), parameter :: CUBE_ROOTS_OF_1 = (/ (1.d0, 0.d0), &
                                                                             exp(complex(0.d0, PIx2/3)), &
                                                                             exp(complex(0.d0, 2*PIx2/3)) /)

CONTAINS

  SUBROUTINE quartic_solve (Poly, out) BIND(C)
    implicit none
    real(c_double), dimension(5), intent(in) :: Poly
    double complex, dimension(4), intent(out) :: out
    real(c_double) :: e, d, c, b, bb, p, q, r, cc, dd, ee, ccc, Delta, Dee, Delta_0, Delta_1
    complex(c_double_complex) :: S, phi, zQ, SSx4, sqrtp, sqrtm
    complex(c_double_complex), dimension(3) :: S_v, SSx4_v
    integer(c_int) :: ir
    e = Poly(1) / Poly(5)
    d = Poly(2) / Poly(5)
    c = Poly(3) / Poly(5)
    b = Poly(4) / Poly(5)
    bb = b*b
    p = 0.125 * (8*c - 3*bb)
    q = 0.125 * (bb*b - 4*b*c + 8*d)
    r = (-3*bb*bb + 256*e - 64*b*d + 16*bb*c) / 256
    c = p
    d = q
    e = r
    cc = c*c
    dd = d*d
    ee = e*e
    ccc = cc*c
    Delta = -27*dd*dd - 128*cc*ee + 256*ee*e + 16*cc*cc*e + 144*c*dd*e - 4*ccc*dd
    Dee = 64*e - 16*cc
    Delta_0 = cc + 12*e
    Delta_1 = 2*ccc + 27*dd - 72*c*e
    if (Delta > 0 .and. p < 0 .and. Dee < 0) then
       phi = acos(Delta_1 / (2*sqrt(complex(Delta_0**3,0.d0))))
       S = 0.5 * sqrt((-2*p + 2*sqrt(complex(Delta_0,0.d0))*cos(phi/3))/3)
       SSx4 = 4*S*S
    else
       zQ = (0.5*(Delta_1 + sqrt(complex(-27*Delta,0.d0))))**(1.d0/3)
       SSx4_v = (-2*p + (zQ*CUBE_ROOTS_OF_1 + Delta_0/(zQ*CUBE_ROOTS_OF_1))) / 3.d0
       S_v = 0.5*sqrt(SSx4_v)
       if (Delta .ne. 0) then
          ir = maxloc(abs(S_v), 1)
          SSx4 = SSx4_v(ir)
          S = S_v(ir)
       else
          SSx4 = SSx4_v(1)
          S = S_v(1)
       endif
    endif
    sqrtp = sqrt(-SSx4 - 2*p + q/S)
    sqrtm = sqrt(-SSx4 - 2*p - q/S)
    out(1) = -0.25*b - S + 0.5*sqrtp;
    out(2) = -0.25*b - S - 0.5*sqrtp;
    out(3) = -0.25*b + S + 0.5*sqrtm;
    out(4) = -0.25*b + S - 0.5*sqrtm;
  END SUBROUTINE quartic_solve

  integer(c_int) function sort_complex_conjugate_pairs(roots) BIND(C)
    implicit none
    complex(c_double_complex), dimension(4), intent(inout) :: roots
    integer(c_int) :: npairs, i, j
    complex(c_double_complex) :: r, r_conj
    npairs = 0
    i = 1
    do while (i <= 3)
       r = roots(i)
       if (abs(DIMAG(r)) > IMAG_TOLERANCE) then
          do j = 1, 4-i
             r_conj = roots(i+j)
             if (       abs(DBLE(r) - DBLE(r_conj)) < ZERO_TOLERANCE &
                  .and. abs(DIMAG(r) + DIMAG(r_conj)) < ZERO_TOLERANCE) then
                roots(i) = roots(2*npairs+1)
                roots(i+j) = roots(2*npairs+2)
                roots(2*npairs+1) = r
                roots(2*npairs+2) = r_conj
                npairs = npairs + 1
                i = i + 1
                exit
             endif
          enddo
       endif
       i = i + 1
    enddo
    sort_complex_conjugate_pairs = npairs
  end function sort_complex_conjugate_pairs

  real(c_double) function find_min_quartic_root_in_real_interval(P, t0, t1) BIND(C)
    implicit none
    real(c_double), dimension(5), intent(in) :: P
    real(c_double), value, intent(in) :: t0, t1
    complex(c_double_complex), dimension(4) :: roots
    integer(c_int) :: npairs, i
    real(c_double) :: min_root
    complex(c_double_complex) :: r
    call quartic_solve(P, roots)
    min_root = huge(1.d0)
    npairs = sort_complex_conjugate_pairs(roots)
    do i = 2*npairs+1, 4
       r = roots(i)
       if (t0 < DBLE(r) .and. DBLE(r) < t1 &
            .and. DBLE(r) < min_root &
            .and. DIMAG(r)**2 / (DBLE(r)**2 + DIMAG(r)**2) < IMAG_TOLERANCE_SQRD) then
          min_root = DBLE(r)
       endif
    enddo
    find_min_quartic_root_in_real_interval = min_root
  end function find_min_quartic_root_in_real_interval

  real(c_double) function find_collision_time(a_i, a_j, R, t0, t1) BIND(C)
    implicit none
    real(c_double), dimension(3,3), intent(in) :: a_i, a_j
    real(c_double), value, intent(in) :: R, t0, t1
    real(c_double), dimension(3,3) :: a_ji
    real(c_double), dimension(0:4) :: P
    real(c_double), dimension(3) :: r_i, r_j, v_i, v_j, omega_i, omega_j
    real(c_double) :: t, v_ijy
    a_ji = a_i - a_j
    p(4) = dot_product(a_ji(:,3), a_ji(:,3))
    p(3) = 2 * dot_product(a_ji(:,3), a_ji(:,2))
    p(2) = dot_product(a_ji(:,2), a_ji(:,2)) + 2 * dot_product(a_ji(:,3), a_ji(:,1))
    p(1) = 2 * dot_product(a_ji(:,2), a_ji(:,1))
    p(0) = dot_product(a_ji(:,1), a_ji(:,1)) - 4*R*R
    ! PRINT *, "P =", P
    t = find_min_quartic_root_in_real_interval(P, t0, t1)
    if (t .ne. huge(1.d0)) then
       r_i = a_i(:,1) + t*a_i(:,2) + t**2*a_i(:,3)
       r_j = a_j(:,1) + t*a_j(:,2) + t**2*a_j(:,3)
       v_i = a_i(:,2) + 2*t*a_i(:,3)
       v_j = a_j(:,2) + 2*t*a_j(:,3)
       v_ijy = dot_product(v_j-v_i, r_j-r_i) / (2*R)
       if (v_ijy > 0) then
          ! PRINT *, "t =", t, " v_ijy =", v_ijy
          find_collision_time = huge(1.d0)
          return
       endif
    endif
    find_collision_time = t
  end function find_collision_time

END MODULE poly_solvers
